# Adquisición de datos

Para poder leer y manipular los datos, primero debemos importar `pandas`, paquete de Python que nos permite manipular datos de forma eficiente y además nos permite leer archivos con extensión `.csv` entre otras.

In [ ]:
import pandas as pd

In [50]:
# URL o PATH, dependiendo el caso, de donde cargamos el archivo a analizar
path_al_csv = 'https://raw.githubusercontent.com/MinervaNunez/Proyecto-de-Python-BEDU/main/RNPEDFC.csv'

# Almacenamos el archivo csv en una estructura de datos para su correcta manipulación y análisis
df = pd.read_csv(path_al_csv, sep = ',', encoding = 'latin-1')

# Análisis exploratorio de datos

Seguimos la metodología vista en clase.

Primero, obtenemos las dimensiones de nuestro `DataFrame`.

In [51]:
df.shape

(36265, 15)

La celda anterior nos dice que nuestro dataframe tiene 36265 renglones (observaciones), y 15 columnas (variables).

Veamos cuáles son nuestras variables.

In [57]:
df.columns

Index(['Fecha en que se le vio por ultima vez',
       'Hora en que se le vio por ultima vez',
       'Pais en que se le vio por ultima vez',
       'Entidad en que se le vio por ultima vez',
       'Municipio en que se le vio por ultima vez',
       'Localidad en que se le vio por ultima vez', 'Nacionalidad', 'Estatura',
       'Complexion', 'Sexo', 'Edad', 'Descripcion de senas particulares',
       'Etnia', 'Discapacidad', 'Dependencia que envio la informacion'],
      dtype='object')

In [64]:
df.dtypes

Fecha en que se le vio por ultima vez        object
Hora en que se le vio por ultima vez         object
Pais en que se le vio por ultima vez         object
Entidad en que se le vio por ultima vez      object
Municipio en que se le vio por ultima vez    object
Localidad en que se le vio por ultima vez    object
Nacionalidad                                 object
Estatura                                     object
Complexion                                   object
Sexo                                         object
Edad                                         object
Descripcion de senas particulares            object
Etnia                                        object
Discapacidad                                 object
Dependencia que envio la informacion         object
dtype: object

De acuerdo con el resultado de la anterior celda, nuestras 15 variables y sus tipos de datos (por defecto) son:
    
| Variable | Tipo |
| -------  | ---- |
| Fecha en que se le vio por ultima vez        | object |
| Hora en que se le vio por ultima vez         | object |
| Pais en que se le vio por ultima vez         | object |
| Entidad en que se le vio por ultima vez      | object |
| Municipio en que se le vio por ultima vez    | object |
| Localidad en que se le vio por ultima vez    | object |
| Nacionalidad                                 | object |
| Estatura                                     | object |
| Complexion                                   | object |
| Sexo                                         | object |
| Edad                                         | object |
| Descripcion de senas particulares            | object |
| Etnia                                        | object |
| Discapacidad                                 | object |
| Dependencia que envio la informacion         | object |


In [65]:
df.index

RangeIndex(start=0, stop=36265, step=1)

Del anterior comando, `pandas` nos dice que nuestras observaciones están indizadas por valores numéricos que empiezan desde el 0 y terminan en 36265.

Finalmente, echamos un vistazo a los datos para darnos una mejor idea de cómo está compuesto nuestro `DataFrame` a analizar.

In [68]:
df.head(10)

Fecha en que se le vio por ultima vez Hora en que se le vio por ultima vez  \
0                            2012-03-18                             05:00:00   
1                            2007-02-05                             02:00:26   
2                            2009-03-14                             08:30:54   
3                            2006-12-23                             08:00:28   
4                            2008-01-11                             12:00:17   
5                            2012-05-27                             12:00:43   
6                            2010-07-05                             11:00:00   
7                            2011-06-13                             01:00:11   
8                            2010-04-24                             01:00:47   
9                            2009-02-09                             02:30:00   

  Pais en que se le vio por ultima vez  \
0                               MEXICO   
1                               MEXICO   
2                               MEXICO   
3                               MEXICO   
4                               MEXICO   
5                               MEXICO   
6                               MEXICO   
7                               MEXICO   
8                               MEXICO   
9                               MEXICO   

  Entidad en que se le vio por ultima vez  \
0                                  SONORA   
1                        ESTADO DE MEXICO   
2                        ESTADO DE MEXICO   
3                        ESTADO DE MEXICO   
4                        ESTADO DE MEXICO   
5                        ESTADO DE MEXICO   
6                               CHIHUAHUA   
7                        ESTADO DE MEXICO   
8                        ESTADO DE MEXICO   
9                                TLAXCALA   

  Municipio en que se le vio por ultima vez  \
0                                   NAVOJOA   
1                            VALLE DE BRAVO   
2                                    CHALCO   
3                       TLALNEPANTLA DE BAZ   
4                                    TOLUCA   
5                            NICOLAS ROMERO   
6                                   BOCOYNA   
7                                 AMECAMECA   
8                                TEOLOYUCAN   
9                                  TLAXCALA   

  Localidad en que se le vio por ultima vez Nacionalidad Estatura Complexion  \
0                                   NAVOJOA     MEXICANA      1.1    DELGADA   
1                            VALLE DE BRAVO     MEXICANA      1.1    DELGADA   
2                                    CHALCO     MEXICANA      1.1    DELGADA   
3                       TLALNEPANTLA DE BAZ     MEXICANA      1.1    DELGADA   
4                                    TOLUCA     MEXICANA      1.1    DELGADA   
5                            NICOLAS ROMERO     MEXICANA      1.1    DELGADA   
6                           CERRO LA VIRGEN     MEXICANA      1.1    DELGADA   
7                                 AMECAMECA     MEXICANA     1.15    ROBUSTA   
8                                TEOLOYUCAN     MEXICANA     1.15    DELGADA   
9                                  TLAXCALA     MEXICANA     1.59    MEDIANA   

     Sexo Edad                  Descripcion de senas particulares  \
0   MUJER    7  TIENE UNA CICATRIZ EN LA RODILLA IZQUIERDA POR...   
1   MUJER    4                                    NO ESPECIFICADO   
2  HOMBRE    6  MANCHAS, PARTE SUPERIOR DE PIERNA DERECHA, CIC...   
3  HOMBRE   10  LUNARES, A LA ALTURA DE LA CINTURAFORMA DE MAN...   
4  HOMBRE    4                                    NO ESPECIFICADO   
5  HOMBRE    9                                    NO ESPECIFICADO   
6  HOMBRE    7                                    NO ESPECIFICADO   
7   MUJER    7  MANCHAS, EN EL TOBILLODERECHA CAUSADA PORQUEMA...   
8  HOMBRE    6               CICATRICES, EN LA CARAVARICELA POCAS   
9   MUJER   13  LUNARES EN CUELLO, EN INDICE IZQ Y ARCO DEL PI...   

   

In [75]:
df.tail(10)

Fecha en que se le vio por ultima vez  \
36255                            2018-04-25   
36256                            2018-04-22   
36257                            2017-07-02   
36258                            2017-04-06   
36259                            2018-04-15   
36260                            2018-04-26   
36261                            2018-01-08   
36262                            2015-02-02   
36263                            2017-08-22   
36264                            2018-04-11   

      Hora en que se le vio por ultima vez  \
36255                             06:30:00   
36256                             11:30:00   
36257                             11:14:59   
36258                             02:00:00   
36259                             06:30:00   
36260                             08:00:00   
36261                             11:00:00   
36262                             07:00:00   
36263                             04:00:00   
36264                             02:00:00   

      Pais en que se le vio por ultima vez  \
36255                               MEXICO   
36256                               MEXICO   
36257                               MEXICO   
36258                               MEXICO   
36259                               MEXICO   
36260                               MEXICO   
36261                               MEXICO   
36262                               MEXICO   
36263                               MEXICO   
36264                               MEXICO   

      Entidad en que se le vio por ultima vez  \
36255                                  PUEBLA   
36256                                  PUEBLA   
36257                                  SONORA   
36258                                  COLIMA   
36259                                  COLIMA   
36260                                  PUEBLA   
36261                                  COLIMA   
36262                                  PUEBLA   
36263                                  COLIMA   
36264                              NUEVO LEON   

      Municipio en que se le vio por ultima vez  \
36255                     SAN MARTIN TEXMELUCAN   
36256                                    PUEBLA   
36257                                   CANANEA   
36258                                MANZANILLO   
36259                                    COLIMA   
36260                                    PUEBLA   
36261                          VILLA DE ALVAREZ   
36262                                     CHILA   
36263                          VILLA DE ALVAREZ   
36264                  SAN NICOLAS DE LOS GARZA   

      Localidad en que se le vio por ultima vez Nacionalidad         Estatura  \
36255                     SAN MARTIN TEXMELUCAN     MEXICANA              1.7   
36256                                    PUEBLA     MEXICANA             1.52   
36257                                   CANANEA     MEXICANA  NO ESPECIFICADO   
36258                                  SANTIAGO     MEXICANA             1.71   
36259                                    COLIMA     MEXICANA             1.59   
36260                                    PUEBLA     MEXICANA              1.6   
36261                          VILLA DE ALVAREZ     MEXICANA              1.7   
36262                      CHILAS DE LAS FLORES     MEXICANA             1.77   
36263                          VILLA DE ALVAREZ     MEXICANA             1.76   
36264                  SAN NICOLAS DE LOS GARZA     MEXICANA              1.6   

            Complexion    Sexo             Edad  \
36255          DELGADA  HOMBRE               26   
36256          DELGADA   MUJER               18   
36257  NO ESPECIFICADO  HOMBRE  NO ESPECIFICADO   
36258          ROBUSTA  HOMBRE               22   
36259          ROBUSTA  HOMBRE               57   
36260          DELGADA  HOMBRE               22   
36261          ROBUSTA   MUJER               27   
36262          MEDIANA   MUJER               86   
36263          ROBUST

Del vistazo a los últimos 10 valores de nuestro dataset podemos observar que nuestro dataset contiene algunos datos faltantes etiquetados por el valor `NO ESPECIFICADO`.

¿Será que existe alguna observación para la cual el pais donde se le vió por última vez sea distinto de México?
Para contestar esta incógnita podemos ver si la cantidad de registros que satisfacen que `df['Pais en que se le vio por ultima vez'] != 'MEXICO']` es mayor a 0

In [79]:
len(df.loc[df['Pais en que se le vio por ultima vez'] != 'MEXICO']) > 0

False

Como la instrucción anterior retornó `False`, entonces el País en el que se le vio por última vez siempre es México.

¿Cuáles son todos los posibles valores que puede tomar la variable `Entidad en que se le vio por ultima vez`?

In [86]:
def val_unicos_col(df, col_name):
    return set(df[col_name])

col_name = 'Entidad en que se le vio por ultima vez'

val_col_entidad = val_unicos_col(df, col_name)
val_col_entidad

{'AGUASCALIENTES',
 'BAJA CALIFORNIA',
 'BAJA CALIFORNIA SUR',
 'CAMPECHE',
 'CHIAPAS',
 'CHIHUAHUA',
 'CIUDAD DE MEXICO',
 'COAHUILA DE ZARAGOZA',
 'COLIMA',
 'DURANGO',
 'ESTADO DE MEXICO',
 'GUANAJUATO',
 'GUERRERO',
 'HIDALGO',
 'JALISCO',
 'MICHOACAN',
 'MORELOS',
 'NAYARIT',
 'NO ESPECIFICADO',
 'NUEVO LEON',
 'OAXACA',
 'PUEBLA',
 'QUERETARO',
 'QUINTANA ROO',
 'SAN LUIS POTOSI',
 'SINALOA',
 'SONORA',
 'TABASCO',
 'TAMAULIPAS',
 'TLAXCALA',
 'VERACRUZ',
 'YUCATAN',
 'ZACATECAS'}

¿Cuáles son todos los posibles valores que puede tomar la variable `Municipio en que se le vio por ultima vez`?

In [88]:
col_name = 'Municipio en que se le vio por ultima vez'

val_col_municipio = val_unicos_col(df, col_name)
val_col_municipio
len(val_col_municipio)

1068

Vemos que son muchos los municipios, por lo que mejor preguntamos por la cantidad de municipios únicos. La respuesta a esta pregunta es 1068 municipios únicos.


¿Cuáles son todos los posibles valores que puede tomar la variable `Localidad en que se le vio por ultima vez`?

In [89]:
col_name = 'Localidad en que se le vio por ultima vez'

val_col_localidad = val_unicos_col(df, col_name)
val_col_localidad
len(val_col_localidad)

2676

De nuevo, vemos que son muchas las localidades, por lo que mejor preguntamos por la cantidad de valores únicos que toma la variable. La respuesta a esta pregunta es 2676 localidades únicas.


Vale la pena preguntarse si la variable `Nacionalidad` toma valores distintos de `MEXICANA`.

In [95]:
col_name = 'Nacionalidad'

val_col_nacionalidad = val_unicos_col(df, col_name)
len(val_col_nacionalidad)

26

Como la cantidad de elementos distintos en la columna es 26, entonces podemos afirmar que en efecto la variable `Nacionalidad` toma valores distintos de `MEXICANA`.

De igual forma podemos investigar qué valores únicos puede tomar la variable `Complexion`.

In [96]:
col_name = 'Complexion'
val_col_complexion = val_unicos_col(df, col_name)
val_col_complexion

{'DELGADA',
 'MEDIANA',
 'NO ESPECIFICADO',
 'No Especificado',
 'OBESA',
 'ROBUSTA'}

De lo anterior podemos observar que en particular en nuestro dataset hay dos formas de especificar que un valor es `NA`:
Puede ser escrito como `"NO ESPECIFICADO"` o como `No Especificado`. Habría que manejar estas dos formas de escribir el valor en la limpieza de los datos.

Podemos preguntarnos si el campo `Sexo` permite el registro de gente no binaria.

In [97]:
col_name = 'Sexo'
val_col_sexo = val_unicos_col(df, col_name)
val_col_sexo

{'HOMBRE', 'MUJER'}

Por lo tanto, la variable `Sexo` sólo admite dos valores: `HOMBRE` y `MUJER`.

Veamos cuál es el rango de edad:

In [119]:
col_name = 'Edad'
val_col_edad = val_unicos_col(df, col_name)
val_col_edad

lista = list(
    map(
        lambda x: int(x) if x.isdigit() else -1, 
        val_col_edad
    )
)
lista.sort()

print(f'El mínimo es {lista[1]} año(s) y el máximo es de {lista[len(lista)-1]} año(s)')

El mínimo es 1 año(s) y el máximo es de 103 año(s)


Entonces, el rango de edad va desde 1 año hasta 103 años.

Ahora investiguemos cuál es el rango de alturas:

In [132]:
def isfloat(s):
    try:
        float(s)
        
        return True
    except:
        return False

col_name = 'Estatura'
val_col_estatura = val_unicos_col(df, col_name)
val_col_estatura

lista = list(
    map(
        lambda x: float(x) if isfloat(x) else -1, 
        val_col_estatura
    )
)
lista.sort()
lista = list(
    filter(lambda x: x != -1, lista)
)

print(f'El mínimo de altura es de {lista[0]} metros y el máximo es de altura es de {lista[len(lista)-1]} metros')

El mínimo de altura es de 0.3 metros y el máximo es de altura es de 170.0 metros


Por lo tanto, el rango de altura es de 0.3 m a 170.0 m.

Cabe destacar que hubieron valores `NO ESPECIFICADO` o `NA`s, por lo que deben limpiarse según convenga.

Los valores que toma la variable `Etnia` son

In [121]:
col_name = 'Etnia'

val_col_etnia = val_unicos_col(df, col_name)
val_col_etnia

{'AMUZGOS OAXACA',
 'CHATINOS',
 'CHICHIMECAS',
 'CHINANTECOS',
 'CHOLES',
 'HUASTECOS',
 'HUICHOLES',
 'MAMES',
 'MAYAS',
 'MEXICANEROS',
 'MIXTECOS',
 'NAHUAS',
 'NO ESPECIFICADO',
 'OTOMIES',
 'PAMES',
 'SERIS',
 'TARAHUMARAS - RARAMURI',
 'TLALPANECOS',
 'TOTONACAS',
 'YAQUIS'}

Otra variable de interés es la variable `Discapacidad`.

In [122]:
col_name = 'Discapacidad'

val_col_discapacidad = val_unicos_col(df, col_name)
val_col_discapacidad

{'AUTISMO', 'NINGUNO', 'NO ESPECIFICADO', 'SINDROME DE DOWN'}

Del bloque anterior podemos apreciar que son 4 posibles valores: `AUTISMO`, `NINGUNO`, `NO ESPECIFICADO` y `SINDROME DE DOWN`

Finalmente, procedemos a investigar la variable `Dependencia que envio la informacion`

In [133]:
col_name = 'Dependencia que envio la informacion'

val_col_dependencia = val_unicos_col(df, col_name)
val_col_dependencia

{'FGE - AGUASCALIENTES',
 'FGE - CAMPECHE',
 'FGE - CHIAPAS',
 'FGE - CHIHUAHUA',
 'FGE - COAHUILA',
 'FGE - DURANGO',
 'FGE - GUERRERO',
 'FGE - JALISCO',
 'FGE - MORELOS',
 'FGE - NAYARIT',
 'FGE - OAXACA',
 'FGE - PUEBLA',
 'FGE - QUERETARO',
 'FGE - QUINTANA ROO',
 'FGE - SAN LUIS POTOSI',
 'FGE - SINALOA',
 'FGE - SONORA',
 'FGE - TABASCO',
 'FGE - VERACRUZ',
 'FGE - YUCATAN',
 'FGE - ZACATECAS',
 'FGJ - ESTADO DE MEXICO',
 'PGJ - BAJA CALIFORNIA',
 'PGJ - BAJA CALIFORNIA SUR',
 'PGJ - CIUDAD DE MEXICO',
 'PGJ - COLIMA',
 'PGJ - GUANAJUATO',
 'PGJ - HIDALGO',
 'PGJ - MICHOACAN',
 'PGJ - NUEVO LEON',
 'PGJ - TAMAULIPAS',
 'PGJ - TLAXCALA'}

# Conclusiones del primer análisis exploratorio de datos

Nuestro dataset está constituido de 36265 observaciones de 15 variables indizadas del 0 al 36264. Nuestras variables son

| Variable |
| -------  |
| Fecha en que se le vio por ultima vez        |
| Hora en que se le vio por ultima vez         |
| Pais en que se le vio por ultima vez         |
| Entidad en que se le vio por ultima vez      |
| Municipio en que se le vio por ultima vez    |
| Localidad en que se le vio por ultima vez    |
| Nacionalidad                                 |
| Estatura                                     |
| Complexion                                   |
| Sexo                                         |
| Edad                                         |
| Descripcion de senas particulares            |
| Etnia                                        |
| Discapacidad                                 |
| Dependencia que envio la informacion         |

De una observación preliminar de los primeros y últimos 10 datos podemos observar que el tipo de dato inferido por pandas es erróneo (el tipo de dato inferido fue `object`).

La propuesta es la siguiente:

| Variable | Tipo |
| -------  | ---- |
| Fecha en que se le vio por ultima vez        | Fecha con formato 'yyyy-mm-dd' |
| Hora en que se le vio por ultima vez         | Hora con formato 'hh:mm:ss' |
| Pais en que se le vio por ultima vez         | Categórico (`MEXICO`) |
| Entidad en que se le vio por ultima vez      | Categórico (valores de `val_col_entidad`) |
| Municipio en que se le vio por ultima vez    | Categórica (valores de `val_col_municipio`) |
| Localidad en que se le vio por ultima vez    | Categórica (valores de `val_col_localidad`) |
| Nacionalidad                                 | Categórica (valores de `val_col_nacionalidad`) |
| Estatura                                     | Numérico float64 (`0.3` a `170.0`) |
| Complexion                                   | Categórica (`DELGADA`, `MEDIANA`, `NO ESPECIFICADA`, `OBESA`, `ROBUSTA`) |
| Sexo                                         | Categórica (`HOMBRE`, `MUJER`) |
| Edad                                         | Numérico int64 (`1` a `103`) |
| Descripcion de senas particulares            | object (string) |
| Etnia                                        | Categórica (valores de  `val_col_etnia`)|
| Discapacidad                                 | Categórica (`AUTISMO`, `NINGUNO`, `NO ESPECIFICADO`, `SINDROME DE DOWN`) |
| Dependencia que envio la informacion         | Categórica (valores de `val_col_dependencia`) |